### PR TITLE
[SYSTEMD] Working systemd unit files

### PR DESCRIPTION
**Pull request by benliao1**
To make `systemd` work properly, we had to split shared memory creation and destruction into two non-persistent processes, `shm_start`, and `shm_stop`. `shm_start` opens and creates all the blocks of shared memory and semaphores needed for shared memory to operate. `shm_start` binds to the existing blocks and opens all the semaphores before closing and unlinking everything, making sure Runtime exits cleanly.

The `systemd` services first start `shm_start` on bootup, wait for it to terminate properly, and then run the three main Runtime processes as background services upon completion of `shm_start`. When any of these three processes fail, the others fail with it and call `shm_stop` to clean up Runtime before restarting all the services again.

Also updated: `shm_client`, `shm_cli`, to reflect new split shared memory processes.

Closes: #41

### CODE REVIEW

**Comment by AVSurfer123** [comment]
@benliao1 can you fix the style by putting common functions in `shm_wrapper.h` first?

**Comment by benliao1** [comment]
@AVSurfer123 Ok, fixed style by moving all semaphore utilities to `shm_wrapper.h` (except `my_sem_unlink` and `my_shm_unlink`, used only by `shm_stop.c`). I didn't fix the style yet / add comments and spaces and stuff--those changes are on `docs_issue_68` so once this goes into `master`, I'll merge `master` into `docs_issue_68` and fix the style there.

File: shm_wrapper/shm_stop.c [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	if ((log_data_shm_ptr = mmap(NULL, sizeof(log_data_shm_t), PROT_READ | PROT_WRITE, MAP_SHARED, fd_shm, 0)) == MAP_FAILED) {
+		log_printf(FATAL, "mmap log_data_shm: %s", strerror(errno));
+		exit(1);
+	}
+	if (close(fd_shm) == -1) {
+		log_printf(ERROR, "close log_data_shm: %s", strerror(errno));
+	}
+    
+     // ******************************* CLOSE + UNLINK EVERYTHING ************************** //
```
Comment (AVSurfer123) [inline shm_wrapper/shm_stop.c:100]: UNLINK + CLOSE

File: shm_wrapper/shm_stop.c [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+		if (sem_unlink((const char *) sname) == -1) {
+			log_printf(ERROR, "sem_unlink: data_sem for dev_ix %d. %s", i, strerror(errno));
+		}
+		generate_sem_name(COMMAND, i, sname);
+		if (sem_unlink((const char *) sname) == -1) {
+			log_printf(ERROR, "sem_unlink: command_sem for dev_ix %d. %s", i, strerror(errno));
+		}
+	}
+	
+	return 1; //returns with 1 to start up
```
Comment (AVSurfer123) [inline shm_wrapper/shm_stop.c:152]: Make comment better to understand why its returning 1

File: tests/client/shm_client.c [hunk 3]
```diff
@@ -16,24 +16,45 @@ void start_shm ()
...
+		if (chdir("../shm_wrapper") == -1) {
+			printf("chdir: %s\n", strerror(errno));
+		}
+
+		//exec the shm_stop process
+		if (execlp("./shm_stop", "shm", (char *) 0) < 0) {
+			printf("execlp: %s\n", strerror(errno));
+		}
+	} else { //parent
+		sleep(1); //allows shm to set itself up
```
Comment (AVSurfer123) [inline tests/client/shm_client.c:52]: Remove this sleep

File: tests/client/shm_client.c [hunk 4]
```diff
@@ -16,24 +16,45 @@ void start_shm ()
...
 		}
 
-		//exec the actual shm process
-		if (execlp("./shm", "shm", (char *) 0) < 0) {
+		//exec the shm_start process
+		if (execlp("./shm_start", "shm", (char *) 0) < 0) {
 			printf("execlp: %s\n", strerror(errno));
 		}
 	} else { //parent
 		sleep(1); //allows shm to set itself up
```
Comment (AVSurfer123) [inline tests/client/shm_client.c:24]: Remove this sleep

**Comment by AVSurfer123** [comment]
@kaqqy Can you update the systemd Wiki with how this cycle works to ensure Runtime is always running?

**Pull request merged**